disnake/auto_moderation.py

```diff
@@ -0,0 +1,495 @@
+
+The above copyright notice and this permission notice shall be included in
+all copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+DEALINGS IN THE SOFTWARE.
+"""
+
+from __future__ import annotations
+
+from datetime import timedelta
+from typing import TYPE_CHECKING, List, Literal, Optional, Sequence, Union, overload
+
+from .enums import (
+    AutomodActionType,
+    AutomodEventType,
+    AutomodKeywordPresetType,
+    AutomodTriggerType,
+    enum_if_int,
+    try_enum,
+    try_enum_to_int,
+)
+from .utils import MISSING, _get_as_snowflake
+
+if TYPE_CHECKING:
+    from typing_extensions import Self
+
+    from .abc import Snowflake
+    from .channel import ForumChannel
+    from .guild import Guild, GuildChannel as GuildChannelType, GuildMessageable
+    from .member import Member
+    from .message import Message
+    from .role import Role
+    from .types.auto_moderation import (
+        AutomodAction as AutomodActionPayload,
+        AutomodActionExecutionEvent as AutomodActionExecutionEventPayload,
+        AutomodActionMetadata,
+        AutomodRule as AutomodRulePayload,
+        AutomodTriggerMetadata as AutomodTriggerMetadataPayload,
+        EditAutomodRule as EditAutomodRulePayload,
+    )
+
+__all__ = ("AutomodAction", "AutomodTriggerMetadata", "AutomodRule", "AutomodActionExecution")
+
+
+class AutomodAction:
+    """
+    Represents an auto moderation action.
+
+    .. versionadded:: 2.6
+
+    Attributes
+    ----------
+    type: :class:`AutomodActionType`
+        The action type.
+    """
+
+    __slots__ = ("type", "_metadata")
+
+    @overload
+    def __init__(self, *, type: Literal[AutomodActionType.block_message]):
+        ...
+
+    @overload
+    def __init__(self, *, type: Literal[AutomodActionType.send_alert_message], channel: Snowflake):
+        ...
+
+    @overload
+    def __init__(
+        self,
+        *,
+        type: Literal[AutomodActionType.timeout],
+        timeout_duration: Union[int, timedelta],
+    ):
+        ...
+
+    def __init__(
+        self,
+        *,
+        type: AutomodActionType,
+        channel: Optional[Snowflake] = None,
+        timeout_duration: Optional[Union[int, timedelta]] = None,
+    ):
+        self.type: AutomodActionType = enum_if_int(AutomodActionType, type)
+        self._metadata: AutomodActionMetadata = {}
+
+        # NOTE: if this is changed to do any sort of processing on those parameters,
+        # `_from_dict` would need to be updated as it doesn't pass any of them
```

Comment:
We're manually assigning `_metadata` in `_from_dict` to allow access to unimplemented fields, not passing them just avoids unnecessary processing, e.g. wrapping the `channel_id` in an `Object` only for the `__init__` to unpack it to int again.
I'm open to alternative ways to do this, but I'd really like to avoid unnecessary overhead where possible :p